tests/aws/app/app_test.go

```diff
@@ -0,0 +1,217 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main_test
+
+import (
+	context "context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/logging/logadmin"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/xray"
+	"github.com/google/go-cloud/tests/internal/testutil"
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/api/iterator"
+)
+
+const (
+	requestlogURL = "/requestlog/"
+	traceURL      = "/trace/"
+)
+
+var (
+	hostIP       string
+	awsRegion    string
+	gcpProjectID string
+	sshUser      string
+	sshKeyPath   string
+)
+
+func init() {
+	flag.StringVar(&hostIP, "host-ip", "", "host IP address of the instance")
+	flag.StringVar(&awsRegion, "aws-region", "us-west-1", "the region used to run the sample app")
+	flag.StringVar(&gcpProjectID, "gcp-project", "", "GCP project used to collect request logs")
+	flag.StringVar(&sshUser, "ssh-user", "admin", "user name used to ssh into the EC2 instance")
+	flag.StringVar(&sshKeyPath, "key-path", "", "path to the key file")
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	if hostIP == "" || gcpProjectID == "" {
+		log.Println("Test environments need to be setup to run server tests.")
+		flag.PrintDefaults()
+		return
+	}
+	if err := startApp(); err != nil {
+		log.Fatal("startApp:", err)
+	}
+	os.Exit(m.Run())
+}
+
+func startApp() error {
+	key, err := publicKeyFile(sshKeyPath)
+	if err != nil {
+		return fmt.Errorf("cannot get key from file: %v", err)
+	}
+	config := &ssh.ClientConfig{
+		User: sshUser,
+		Auth: []ssh.AuthMethod{key},
+		HostKeyCallback: func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+			return nil
+		},
+		Timeout: 30 * time.Second,
+	}
+	c, err := ssh.Dial("tcp", hostIP+":22", config)
+	if err != nil {
+		return fmt.Errorf("failed to dial %s:22: %v", hostIP, err)
+	}
+	session, err := c.NewSession()
+	if err != nil {
+		return fmt.Errorf("failed to create session: %v", err)
+	}
+
+	// This close the session channel and the TCP connection, but doesn't terminate
+	// the running app. session.Signal() does not terminate it either, see
+	// https://github.com/golang/go/issues/4115. This is OK for our use case since
+	// we tear it down after the test finishes.
+	defer func() {
+		session.Close()
+		c.Close()
+	}()
+	return session.Start("/home/admin/app 2>&1 | logger")
+}
+
+func publicKeyFile(path string) (ssh.AuthMethod, error) {
```

Comment:
This function needs a more clear name, as the path you are giving is for a private key file. 🙂

Reply:
Split the PublicKeys part out, renamed to signerFromFile.